icu4j/main/classes/core/src/com/ibm/icu/impl/units/UnitsData.java

```diff
@@ -168,7 +168,7 @@ public CategoriesSink() {
 
         @Override
         public void put(UResource.Key key, UResource.Value value, boolean noFallback) {
-            assert (key.toString() == Constants.CATEGORY_TABLE_NAME);
```

Comment:
This does not currently cause assertion failure. As long as the String is interned, `==` operation works. However, it looks `toString()` here seems to call `substring()`, which may create a new String instance.

Anyway, I don't think this assertion is not expecting String object's identity. We may hit this assertion unexpectedly with `==` through a certain code path, or future code changes. I believe this assertion is for String value equality, so we should use `equals()` method here. Again, static code analysis tools don't like such code.

Reply:
Yeah, good catch; I didn't see when reviewing this code that these were strings.